graph/graph.go

```diff
@@ -292,6 +308,11 @@ func (graph *Graph) Register(img *image.Image, layerData io.Reader) (err error)
 		return err
 	}
 	graph.idIndex.Add(img.ID)
+
+	graph.imageMutex.Lock(img.Parent)
+	graph.parentRefs[img.Parent]++
+	graph.imageMutex.Unlock(img.Parent)
+
```

Comment:
I think there is a possible deadlock here given that `attemptIDReuse` (see https://github.com/docker/docker/pull/16952/files#diff-7e2a216082b6bd557c7ecebd68305e2eR576) may lock images in a different order.

cc @tonistiigi